src/gf_ring.h

```diff
@@ -520,8 +532,17 @@ void RingModN<T>::compute_omegas_cached(vec::Vector<T>& W, int n, T w) const
 template <typename T>
 T RingModN<T>::rand(void) const
 {
-    std::uniform_int_distribution<uint32_t> dis(1, this->card() - 1);
-    return dis(prng());
+    return distribution(prng());
+}
+
+template <>
+inline __uint128_t RingModN<__uint128_t>::rand(void) const
+{
+    __uint128_t c = distribution(prng());
+    if (_card > std::numeric_limits<uint64_t>::max()) {
+        c |= ((((distribution_leading(prng()) << 16) << 16) << 16) << 16);
```

Comment:
could be simplified as `c |= (static_cast<__uint128_t>(distribution(prng())) << 64);`